src/OpenFeature/Api.cs

```diff
@@ -38,25 +39,48 @@ public sealed class Api : IEventBus
         private Api() { }
 
         /// <summary>
-        /// Sets the feature provider. In order to wait for the provider to be set, and initialization to complete,
+        /// Sets the default feature provider to given clientName without awaiting its initialization.
+        /// </summary>
+        /// <remarks>The provider cannot be set to null. Attempting to set the provider to null has no effect.</remarks>
+        /// <param name="featureProvider">Implementation of <see cref="FeatureProvider"/></param>
+        [Obsolete("Will be removed in later versions; use SetProviderAsync, which can be awaited")]
+        public void SetProvider(FeatureProvider featureProvider)
+        {
+            this.EventExecutor.RegisterDefaultFeatureProvider(featureProvider);
+            this._repository.SetProvider(featureProvider, this.GetContext()).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Sets the default feature provider. In order to wait for the provider to be set, and initialization to complete,
         /// await the returned task.
         /// </summary>
         /// <remarks>The provider cannot be set to null. Attempting to set the provider to null has no effect.</remarks>
         /// <param name="featureProvider">Implementation of <see cref="FeatureProvider"/></param>
-        public async Task SetProvider(FeatureProvider featureProvider)
+        public async Task SetProviderAsync(FeatureProvider featureProvider)
         {
             this.EventExecutor.RegisterDefaultFeatureProvider(featureProvider);
             await this._repository.SetProvider(featureProvider, this.GetContext()).ConfigureAwait(false);
         }
 
+        /// <summary>
+        /// Sets the feature provider to given clientName without awaiting its initialization.

```

Comment:
> I feel like it would be safer/more consistent with prior behavior to block on these rather than let the tasks go unobserved.

I'm not sure I agree. The purpose of the SDK's events is to be able to attach handlers, one of which is "PROVIDER_READY". Some implementations have non blocking provider settings, and then use an event to understand when to evaluate flags.

Reply:
Okay, that's interesting, and sounds like I may have misunderstood some intent with the async init changes (still wrapping my head around the wider spec).

Reframing the scenario I was thinking about before:

In #181, I'm registering an implementation of `IHostedLifecycleService` with the idea being to:
1. resolve:
   - `Api` from the container (i.e. currently a singleton registration of `Api.Instance`)
   - `IEnumerable<FeatureProvider>` (i.e. all registrations of type `FeatureProvider`)
2. perform async setup on the `Api` instance when the `IHost` calls `IHostedLifecycleService.StartingAsync(...)` which is the earliest hook we have inside of the `IHost` (i.e. before any calls are made to any `IHostedService.StartAsync(...)`)

Then as a consumer of `OpenFeature.Extensions.Hosting`, I can inject `IFeatureClient` into some `IHostedService` and feel confident knowing that the SDK is fully configured by the time the `IHost` calls my `IHostedService.StartAsync(...)` method.

Now this is somewhat moot for #181 since it has access to the new `Api.SetProviderAsync(...)` method and can thus choose to wait for the async init to complete, but #181 is based on some internal code I wrote, and if this change were to ship _without_ blocking, I have this itchy feeling that my first `IHostedService` could (theoretically?) end up using its injected `IFeatureClient` before the intended `FeatureProvider` has been added to the underlying `ProviderRepository._featureProviders`.

Again, I'm not arguing in favor of blocking (especially if the spec is ambivalent), just thinking out loud about how this _might_ surprise me in my existing code if I weren't aware of the change.

Given that our plan is to ship this sync-over-unawaited-async method at the same time as we're exposing the new async method, I'll just plan to migrate my existing usages to the async overloads when I upgrade (barring #181 also shipping at the same time), so no big deal, but that itchy feeling makes me think that we might want to call this out as a _we-dont-consider-this-breaking-but-take-note-before-you-ugrade_ in the release notes.

Or maybe at least suggest to contrib providers that if/when they opt-into async init, that they call it out as a breaking change in _their_ release notes. _(The `dotnet/efcore` team uses a `providers-beware` label for things like this where the breaking change may or may not impact consumers based on how the provider uses the feature before/after the change in question.)_

### Related snippets

https://github.com/open-feature/dotnet-sdk/blob/1efe09da3948d5dfd7fd9f1c7a040fc5c2cbe833/src/OpenFeature/OpenFeatureClient.cs#L39-L48

https://github.com/open-feature/dotnet-sdk/blob/1efe09da3948d5dfd7fd9f1c7a040fc5c2cbe833/src/OpenFeature/ProviderRepository.cs#L251-L261

https://github.com/open-feature/dotnet-sdk/blob/2eb7335eb3843de93dd5b0b05745ff0eacf58c16/src/OpenFeature.Extensions.Hosting/Internal/OpenFeatureHostedService.cs#L11-L36

Reply:
> Given that our plan is to ship this sync-over-unawaited-async method at the same time as we're exposing the new async method, I'll just plan to migrate my existing usages to the async overloads when I upgrade (barring https://github.com/open-feature/dotnet-sdk/pull/181 also shipping at the same time), **_so no big deal, but that itchy feeling makes me think that we might want to call this out as a we-dont-consider-this-breaking-but-take-note-before-you-ugrade in the release notes._**

Yes, absolutely. The thing to note though that might make you feel better, is that since we haven't released init/shutdown yet, there really aren't semantics around provider readiness in the wild yet in .NET (it's conformant to a very old OpenFeature spec); there's basically no existing provider lifecycle defined. As of now, providers are for all practical purposes assumed to be ready when you set them.

Or maybe at least suggest to contrib providers that if/when they opt-into async init, that they call it out as a breaking change in their release notes. (The dotnet/efcore team uses a providers-beware label for things like this where the breaking change may or may not impact consumers based on how the provider uses the feature before/after the change in question.)

100%.

I would suggest that all providers in our contribs would be upgraded to require at minimum this new release of the SDK, and consider that a breaking change so that it's captured in their release notes. We could do that with a single PR in the contribs (providers hosted elsewhere should take note, and we can reach out to those responsible for them).